odxtools/decodestate.py

```diff
@@ -87,32 +93,125 @@ def extract_atomic_value(
+        raw_value, = bitstruct.unpack_from(
             f"{base_data_type.bitstruct_format_letter}{bit_length}",
             extracted_bytes,
             offset=padding)
-
-        text_errors = 'strict' if exceptions.strict_mode else 'replace'
-        if base_data_type == DataType.A_ASCIISTRING:
-            assert isinstance(internal_value, (bytes, bytearray))
-            # The spec says ASCII, meaning only byte values 0-127.
-            # But in practice, vendors use iso-8859-1, aka latin-1
-            # reason being iso-8859-1 never fails since it has a valid
-            # character mapping for every possible byte sequence.
-            text_encoding = 'iso-8859-1'
-            internal_value = internal_value.decode(encoding=text_encoding, errors=text_errors)
-        elif base_data_type == DataType.A_UTF8STRING:
-            assert isinstance(internal_value, (bytes, bytearray))
-            text_encoding = "utf-8"
-            internal_value = internal_value.decode(encoding=text_encoding, errors=text_errors)
-        elif base_data_type == DataType.A_UNICODE2STRING:
-            assert isinstance(internal_value, (bytes, bytearray))
-            # For UTF-16, we need to manually decode the extracted
-            # bytes to a string
-            text_encoding = "utf-16-be" if is_highlow_byte_order else "utf-16-le"
-            internal_value = internal_value.decode(encoding=text_encoding, errors=text_errors)
+        internal_value: AtomicOdxType
+
+        # Deal with raw byte fields, ...
+        if base_data_type == DataType.A_BYTEFIELD:
+            odxassert(base_type_encoding is None or base_type_encoding == Encoding.NONE)
+
+            internal_value = raw_value
+
+        # ... string types, ...
+        elif base_data_type in (DataType.A_UTF8STRING, DataType.A_ASCIISTRING,
+                                DataType.A_UNICODE2STRING):
+            text_errors = 'strict' if strict_mode else 'replace'
+            if base_type_encoding == Encoding.UTF8:
+                internal_value = raw_value.decode("utf-8", errors=text_errors)
+            elif base_type_encoding == Encoding.UCS2:
+                text_encoding = "utf-16-be" if is_highlow_byte_order else "utf-16-le"
+                internal_value = raw_value.decode(text_encoding, errors=text_errors)
+            elif base_type_encoding == Encoding.ISO_8859_1:
+                internal_value = raw_value.decode("iso-8859-1", errors=text_errors)
+            elif base_type_encoding == Encoding.ISO_8859_2:
+                internal_value = raw_value.decode("iso-8859-2", errors=text_errors)
+            elif base_type_encoding == Encoding.WINDOWS_1252:
+                internal_value = raw_value.decode("cp1252", errors=text_errors)
+            else:
+                odxassert(
+                    base_type_encoding in (None, Encoding.NONE),
+                    f"Specified illegal encoding {base_type_encoding} for string object")
```

Comment:
Specs explicitly forbid this?

Reply:
good point. In my data sets this is the case, but after checking the spec, the default encoding for strings depends on the base data type (cf table 8) and so do the default encodings of integers. fixed: https://github.com/mercedes-benz/odxtools/pull/381/commits/c84241bad312d0e59f8d63060ba8e2a1a0178c27